File: rti_connext_dds-5.2.3/resource/template/rti_workspace/examples/connext_dds/c++/hello_dynamic/src/HelloWorldType.cxx
/*****************************************************************************/
/*         (c) Copyright, Real-Time Innovations, All rights reserved.        */
/*                                                                           */
/*         Permission to modify and use for internal purposes granted.       */
/* This software is provided "as is", without warranty, express or implied.  */
/*                                                                           */
/*****************************************************************************/

#include <iostream>
#include "dds_cpp/dds_cpp_domain.h"
#include "ndds/ndds_cpp.h"
#include "HelloWorldType.h"
#include "Hello.h"  // Contains the constants HELLODDS_MAX_PAYLOAD_SIZE
                    // and HELLODDS_MAX_STRING_SIZE.


/*****************************************************************************/
/* HelloWorldType_get_type_name                                              */
/*                                                                           */
/* Returns a C String with the name of this data type                        */
/*****************************************************************************/
const char *HelloWorldType_get_type_name() {
    return "HelloWorld";
}



/*****************************************************************************/
/* HelloWorldType_create                                                     */
/*                                                                           */
/* Creates type code for a structure corresponding to:                       */
/*                                                                           */
/* struct HelloWorld {                                                       */
/*    string<HELLODDS_MAX_STRING_SIZE>             prefix;                   */
/*    long                                         sampleId;                 */
/*    sequence<octet, HELLODDS_MAX_PAYLOAD_SIZE>   payload;                  */
/* };                                                                        */
/*                                                                           */
/* Returns NULL in case of error                                             */
/*****************************************************************************/
DDS_TypeCode *HelloWorldType_create() {
    DDS_TypeCodeFactory * factory = NULL;
    DDS_TypeCode * sequenceTc = NULL;   // Typecode for the 'payload' sequence
    DDS_TypeCode * stringTc = NULL;     // Typecode for the 'prefix' string;
    DDS_TypeCode * structTc = NULL;     // Top-level typecode
    DDS_ExceptionCode_t ex;

    factory = DDS_TypeCodeFactory::get_instance();
    if (factory == NULL) {
        std::cerr << "! Unable to get type code factory singleton" 
                  << std::endl;
        goto done;
    }
    
    // Create typecode for 'prefix' string of max size HELLODDS_MAX_STRING_SIZE
    stringTc = factory->create_string_tc(HELLODDS_MAX_STRING_SIZE, ex);
    if (ex != DDS_NO_EXCEPTION_CODE) {
        std::cerr << "! Unable to create 'prefix' string typecode: " << ex 
                  << std::endl;
        goto done;
    }
    
    // Create typecode for 'payload' octet sequence
    sequenceTc = factory->create_sequence_tc(HELLODDS_MAX_PAYLOAD_SIZE,
                        factory->get_primitive_tc(DDS_TK_OCTET),
                        ex);
    if (ex != DDS_NO_EXCEPTION_CODE) {
        std::cerr << "! Unable to create 'payload' sequence typecode: " << ex
                  << std::endl;
        goto done;
    }
    structTc = factory->create_struct_tc("HelloWorld", DDS_StructMemberSeq(), 
                        ex);
    if (ex != DDS_NO_EXCEPTION_CODE) {
        std::cerr << "! Unable to create struct typecode, error=" << ex 
                  << std::endl;
        goto done;
    }
    structTc->add_member("prefix",
                        DDS_MEMBER_ID_INVALID,
                        stringTc,
                        DDS_TYPECODE_NONKEY_MEMBER,
                        ex);
    if (ex != DDS_NO_EXCEPTION_CODE) {
        std::cerr << "! Unable to add member to struct typecode, error=" << ex 
                  << std::endl;
        goto done;
    }

    structTc->add_member("sampleId",
                        DDS_MEMBER_ID_INVALID,
                        factory->get_primitive_tc(DDS_TK_LONG),
                        DDS_TYPECODE_NONKEY_MEMBER,
                        ex);
    if (ex != DDS_NO_EXCEPTION_CODE) {
        std::cerr << "! Unable to add member to struct typecode, error=" << ex 
                  << std::endl;
        goto done;
    }
    
    structTc->add_member("payload",
                        DDS_MEMBER_ID_INVALID,
                        sequenceTc,
                        DDS_TYPECODE_NONKEY_MEMBER,
                        ex);
    if (ex != DDS_NO_EXCEPTION_CODE) {
        std::cerr << "! Unable to add member to struct typecode, error=" << ex 
                  << std::endl;
        goto done;
    }

    if (sequenceTc != NULL) {
        factory->delete_tc(sequenceTc, ex);
        sequenceTc = NULL;
    }
    if (stringTc != NULL) {
        factory->delete_tc(stringTc, ex);
        stringTc = NULL;
    }

    return structTc;

done:
    // In case of failure, do a cleanup of all the temporary data structures
    if (sequenceTc != NULL) {
        factory->delete_tc(sequenceTc, ex);
    }
    if (stringTc != NULL) {
        factory->delete_tc(stringTc, ex);
    }
    if (structTc != NULL) {
        factory->delete_tc(structTc, ex);
    }
    return NULL;
}

/*****************************************************************************/
/* HelloWorldType_delete                                                     */
/*                                                                           */
/* Deletes type code created by HellOWorldType_create                        */
/*                                                                           */
/*****************************************************************************/
void HelloWorldType_delete(DDS_TypeCode *type) {
    DDS_TypeCodeFactory * factory = NULL;
    DDS_ExceptionCode_t ex;

    factory = DDS_TypeCodeFactory::get_instance();
    if (factory == NULL) {
        std::cerr << "! Unable to get type code factory singleton" 
                  << std::endl;
        return;
    }

    if (type != NULL) {
        factory->delete_tc(type, ex);
    }
    if (ex != DDS_NO_EXCEPTION_CODE) {
        std::cerr << "! Unable to delete typecode: " << ex 
                  << std::endl;
        return;
    }

}

File: rti_connext_dds-5.2.3/resource/template/rti_workspace/examples/connext_dds/cs/hello_dynamic/src/HelloWorldType.cs
/* ****************************************************************************
*         (c) Copyright, Real-Time Innovations, All rights reserved.       
*                                                                          
*         Permission to modify and use for internal purposes granted.      
* This software is provided "as is", without warranty, express or implied. 
*                                                                          
* ****************************************************************************
*/
using System;
using System.Collections.Generic;
using System.Text;

namespace Hello
{
    /**
     * A simple class containing utility methods (static) to create HelloWorld
     * data type dynamically. 
     */
    class HelloWorldType
    {
        public const int HELLODDS_MAX_STRING_SIZE = 64;
        public const int HELLODDS_MAX_PAYLOAD_SIZE = 8192;

        /**************************************************************************
         * Returns the name of the type
         */
        public static String getTypeName() {
            return "HelloWorld";
        }

        /**************************************************************************
         * Creates type code for a structure corresponding to:
         *
         * struct HelloWorld {
         *    string<HELLODDS_MAX_STRING_SIZE>             prefix;
         *    long                                         sampleId;
         *    sequence<octet, HELLODDS_MAX_PAYLOAD_SIZE>   payload;
         * };
         *
         * Returns NULL in case of error
         */
        public static DDS.TypeCode create()
        {
            DDS.TypeCodeFactory factory = DDS.TypeCodeFactory.get_instance();
            if (factory == null)
            {
                Console.Error.WriteLine("! Unable to get type code factory singleton");
                return null;
            }
            /* Create typecode for 'prefix' string of max size 
             * HELLODDS_MAX_STRING_SIZE
             */
            DDS.TypeCode stringTc = null;
            try
            {
                stringTc = factory.create_string_tc(HELLODDS_MAX_STRING_SIZE);
            }
            catch (DDS.ExceptionCode_BadParam err)
            {
                Console.Error.WriteLine("! Unable to create 'prefix' string typecode: " +
                    err.Message);
                return null;
            }
            /* Create typecode for 'payload' octet sequence */
            DDS.TypeCode sequenceTc = null;
            try
            {
                sequenceTc = factory.create_sequence_tc(HELLODDS_MAX_PAYLOAD_SIZE,
                        factory.get_primitive_tc(DDS.TCKind.TK_OCTET));
            }
            catch(DDS.ExceptionCode_BadParam err) {
                Console.Error.WriteLine("! Unable to create 'payload' sequence typecode: " +
                    err.Message);
                return null;
            }
            /* Populate the members of the top-level structure */
            DDS.StructMemberSeq members = new DDS.StructMemberSeq();
            members.ensure_length(3, 3);

            members.buffer[0] = new DDS.StructMember();
            members.buffer[0].name = "prefix";
            members.buffer[0].is_pointer = false;
            members.buffer[0].bits = (short)-1;
            members.buffer[0].is_key = false;
            members.buffer[0].type = stringTc;
            
            members.buffer[1] = new DDS.StructMember();
            members.buffer[1].name = "sampleId";
            members.buffer[1].is_pointer = false;
            members.buffer[1].bits = (short)-1;
            members.buffer[1].is_key = false;
            members.buffer[1].type = factory.get_primitive_tc(DDS.TCKind.TK_LONG);

            members.buffer[2] = new DDS.StructMember();
            members.buffer[2].name = "payload";
            members.buffer[2].is_pointer = false;
            members.buffer[2].bits = (short)-1;
            members.buffer[2].is_key = false;
            members.buffer[2].type = sequenceTc;

            DDS.TypeCode structTc = null;
            try {
                structTc = factory.create_struct_tc("HelloWorld", members);
            }
            catch(DDS.ExceptionCode_BadParam err) {
                Console.Error.WriteLine("! Unable to create struct typecode: " 
                    + err.Message);
                return null;
            }
            return structTc;                
        }

        public static void delete(DDS.TypeCode type) {
            if (type != null) {
                DDS.TypeCodeFactory.get_instance().delete_tc(type);
            }
        }
    }
}
